File: Homework/Ngan_Luu/1372-longest-zigzag-path-in-a-binary-tree.py
# 1372. https://leetcode.com/problems/longest-zigzag-path-in-a-binary-tree/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(node,prevDirection):
            if not(node):
                return -1
            left=dfs(node.left,'left')
            right=dfs(node.right,'right')
            longestZigZag[0]=max(longestZigZag[0],max(left,right)+1)
            if prevDirection=='left':
                return right+1
            else:
                return left+1
        longestZigZag=[0]
        dfs(root,'right')
        return longestZigZag[0]

File: Homework/Duc_Trinh/1372-Longest-ZigZag-Path-in-a-Binary-Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

 const getLongestZigZag = (node, dir, answer) => {
    if (!node) return 0
    const left = getLongestZigZag(node.left, 'left', answer)
    const right = getLongestZigZag(node.right, 'right', answer)
    answer.longest = Math.max(answer.longest, Math.max(left, right) )
    return dir === 'left' ? 1 + right : 1 + left
}
var longestZigZag = function(root) {
  const answer = {
      longest: 0
  }
  if (root.left) {getLongestZigZag(root, 'left', answer)}
  if (root.right) {getLongestZigZag(root, 'right', answer)}
  return answer.longest
};
